cxx-qt-lib/src/qcolor.rs

```diff
@@ -60,153 +60,175 @@ pub struct QColor {
-/// /*???*/;`. The resulting [`Pin`] can be deref'd to `&QColor` as needed.
-///
-/// # Example
-///
-/// ```ignore
-/// use cxx_qt_lib::{let_qcolor, QColor};
-///
-///
-/// fn f(s: &QColor) {/* ... */}
-///
-/// fn main() {
-///     let_qcolor!(s = "example");
-///     f(&s);
-/// }
-/// ```
-#[macro_export]
-macro_rules! let_qcolor {
-    ($var:ident = $value:expr $(,)?) => {
-        let mut stack_qcolor = $crate::private::StackQColor::new();
-        #[allow(unused_mut, unused_unsafe)]
-        let mut $var = match $value {
-            let_qcolor => unsafe { stack_qcolor.init(let_qcolor) },
-        };
-    };
-}
-
 impl QColor {
     /// `QColor` is not constructible via `new`.
-    /// Instead, use the [`let_qcolor!`] macro.
+    /// Instead, use [`Color::from_rgba`].
+    //
+    // FIXME: Do we still need this?
     pub fn new<T: Private>() -> Self {
         unreachable!()
     }
 
     /// Create a new Rust Color from this QColor.
     /// This is a copy operation so any changes will not propagate to the original QColor.
-    pub fn to_rust(&self) -> Option<Color> {
-        // Given that a QColor can only be constructed using [`let_qcolor!`] macro,
-        // it is safe to assume that self is a valid QColor reference which makes these
-        // function calls safe.
-        match unsafe { qcolor_get_spec(self) } {
-            QColorSpec::Unsupported => None,
-            QColorSpec::Rgb => Some(Color::ARGB {
-                alpha: unsafe { qcolor_get_alpha(self) },
-                red: unsafe { qcolor_get_red(self) },
-                green: unsafe { qcolor_get_green(self) },
-                blue: unsafe { qcolor_get_blue(self) },
-            }),
-        }
+    pub fn to_rust(&self) -> Color {
+        Color::from_qcolor(self)
     }
 }
 
-#[doc(hidden)]
-#[repr(C)]
-pub struct StackQColor {
-    // Static assertions in cxx_qt.cpp validate that this
-    // is large enough and aligned enough.
-    space: MaybeUninit<[usize; 2]>,
+// Safety:
+//
+// The code in this file ensures that QColor can only ever be allocated
+// on the stack in pinned form which avoids the pitfalls of trying to
+// move this type that has a non-trivial move constructor.
+unsafe impl ExternType for QColor {
+    type Id = type_id!("QColor");
+    type Kind = cxx::kind::Opaque;
 }
 
-#[allow(missing_docs)]
-impl StackQColor {
-    pub fn new() -> Self {
-        StackQColor {
-            space: MaybeUninit::uninit(),
+/// The Rust representation of Qt's QColor
+///
+/// Internally this holds a UniquePtr to a QColor which has been constructed on the C++ side.
+pub struct Color {
+    // Note that once map_qt_value is removed later, this can become private again
+    #[doc(hidden)]
+    pub inner: cxx::UniquePtr<QColor>,
+}
+
+impl Color {
+    /// Construct a Rust Color from an existing QColor, this is a copy operation.
+    pub fn from_qcolor(qcolor: &QColor) -> Self {
+        Self {
+            // Safety: TODO
```

Comment:
For these `Safety: TODO` comments here, we can probably link to the design documents in the book that we're planning to add, once that is available.

Reply:
Right :-) I felt like i was going to need to write the same statement here many times, linking to the book is a good idea later so lets do that.